challenges/2021-04-13-single-and-ready-to-mingle/solutions/ruby/ericwburden/ericwburden.rb

```diff
@@ -0,0 +1,45 @@
+# # Single and Ready to Mingle
+# 
+# Given an array of integers in which two elements appear exactly once and all other
+# elements appear exactly twice, find the two elements that appear only once.
+# 
+# ## Business Rules/Errata
+# 
+# - You can assume that the input will **always** include at least two and exactly two
+#   elements that appear only once.
+# - You may not assume that the list will be sorted ahead of time.
+# - Extra Challenge: Can you complete this puzzle in linear time and constant space?
+# - Extra Extra Challenge: Can you make your solution generic over other input types?
+# 
+# ## Examples
+# 
+# ```
+# input = [2, 4, 6, 8, 10, 2, 6, 10];
+# find_singles(input)  // [4, 8]
+# ```
+# ```
+# input = [1, 2, 3, 2, 1, 4, 4, 6, 7, 8, 7, 6, 9, 9];
+# find_singles(input)  // [3, 8]
+# ```
+
+require 'set'
+
+def find_singles(numbers)
+  if !numbers.all? {|x| x.is_a?(Integer) } or !numbers.kind_of?(Array)
+    raise "Tried to pass something besides an array of numbers"
+  end
+
+  number_set = Set.new()
+  numbers.each do |item|
```

Comment:
Above, you check for `all?` which iterates through the entire array, perhaps you would check each `item` here as being an `instance_of?` since you're already iterating.

In fact, this plays into ruby's power a bit. You want to raise an error above if the argument isn't an array, and also you want to raise an error here if the specific item isn't an integer.


Oftentimes in these cases we'd extract error handling out into it's own method.  

```
def throw_incorrect_argument_error
  raise 'Tried to pass something besides an array of numbers'
end  
```


Then you can use ruby's online syntax:


`throw_incorrect_argment_error unless numbers.instance_of?(Array)`


Then inside your `.each`

`throw_incorrect_argument_error unless item.instance_of?(Integer)`

Reply:
This really is nitpicking style, but often times we want our code to read like a sentence. 


That _should_ lead to longer more descriptive method names, and for the purposes of an algorithm challenge isn't necessary.

Reply:
That makes sense, and it's definitely easier to read than what I wrote.